### PR TITLE
Service skipchain

### PR DESCRIPTION
**Pull request by ineiti**
Re-opened against master
medco is failing - not related to skipchain

### CODE REVIEW

File: services/skipchain/msgs.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		// Propagation
+		&PropagateSkipBlock{},
+		// Requests for data
+		&GetUpdateChain{},
+		&GetUpdateChainReply{},
+		// Data-structures
+		&ForwardSignature{},
+		&SkipBlockFix{},
+		&SkipBlock{},
+	}
```
Comment (Liamsi) [inline services/skipchain/msgs.go]: Didn't we already had and reviewd that? I remember we coded that together before OSDI?

File: services/skipchain/skipchain.go [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+// If the the latest block given is nil it verify if we are actually creating
+// the first (genesis) block and creates it. If it is called with nil although
+// there already exist previous blocks, it will return an error.
+func (s *Service) ProposeSkipBlock(e *network.ServerIdentity, psbd *ProposeSkipBlock) (network.Body, error) {
+	prop := psbd.Proposed
+	var prev *SkipBlock
+
+	if !psbd.LatestID.IsNull() {
+		// We're appending a block to an existing chain
+		var ok bool
```
Comment (Liamsi) [inline services/skipchain/skipchain.go]: why not use `ok :=`

Comment (Liamsi) [inline services/skipchain/skipchain.go]: Ah I see, because of prev

File: services/skipchain/msgs.go (hunk 1 above)
Comment (ineiti) [inline services/skipchain/msgs.go]: Never mad it into the R4M queue I believe :(

File: services/skipchain/struct.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+		return err
+	}
+	//for _, fl := range sb.ForwardLink {
+	//	if err := fl.VerifySignature(sb.Aggregate); err != nil {
+	//		return err
+	//	}
+	//}
+	//if sb.ChildSL != nil && sb.ChildSL.Hash == nil {
+	//	return sb.ChildSL.VerifySignature(sb.Aggregate)
+	//}
```
Comment (ineiti) [inline services/skipchain/struct.go]: Yes, #478.

**Comment by ineiti** [comment]
AFAIR the OSDI-part was hacked together and did not work at all as it was never used. This is why it hasn't been merged ;) Now it's cleaned-up (remember the structure/interface disaster we had?), uses BFTCoSi and 'only' needs to pass Travis and check all new blocks.

**Comment by nikkolasg** [comment]
OK

**Pull request merged**